rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -321,224 +308,227 @@ object ExecutionBatch extends Loggable {
+                  , Pending(currentConfig, None, expireTime)
+                  )
+                }
+              }
+          }
 
-              if(expireTime.isBefore(now)) {
-                runType("no run (ever or too old)", NoReportInInterval(currentConfig))
+        //
+        // [II] Second case: we DO have a run.
+        // We need to look if it has an associated expected configuration,
+        // and if not try to gather piece of information from elsewhere.
+        // Then analyse the consistancy of the result.
+        //
+        case Some(runInfos) =>
+//          ComplianceDebugLogger.node(nodeId).debug(s"Node run configuration: ${(nodeId, complianceMode, runInfos).toLog }")
+//
+//          val computed =    computeNodeRunInfo(
+//                    nodeId, optInfo, missingReportType
+//                  , intervalInfo, updateValidityTime(intervalInfo), runValidityTime(intervalInfo)
+//                  , now, run
+//              )
+
+          (runInfos, currentNodeConfigs.get(nodeId).flatten) match {
+
+            //
+            // #1 : What the hell ?
+            //      that's not good. Why a run without expected config ?
+            //      More over, we group together the cases where we have config for the run, because without
+            //      a current expected config, it should be impossible (no way to paired it with that run).
+            //      Just log an error for dev.
+            case ((AgentRunWithNodeConfig(AgentRunId(_, t), optConfigId, _, _)), None) =>
+              if(nodeConfigIdInfos.isDefinedAt(nodeId)) {
+                runType("nodeId exists in DB but has no version (due to cleaning?). Need regeneration, no expected report yet.", NoExpectedReport(t, None))
               } else {
-                runType(s"no run (ever or too old), Pending until ${expireTime}", Pending(currentConfig, None, expireTime, missingReportType))
+                runType("nodeId was not found in DB but is sending reports. It is likely a new node. Need regeneration, no expected report yet."
+                , NoExpectedReport(t, None)
+                )
               }
-            }
 
-          case (Some(run), _) =>
-            computeNodeRunInfo(
-                  nodeId, optInfo, missingReportType
-                , intervalInfo, updateValidityTime(intervalInfo), runValidityTime(intervalInfo)
-                , now, run
-            )
-
-        }
-      }
-
-      (nodeId, runInfo)
-    }.toMap
-  }
+            //
+            // #2 : run without config ID (neither in it nor found)
+            //      no expected config for the run. Why so? At least, a recent config.
+            case ((AgentRunWithNodeConfig(AgentRunId(_, t), None, _, _)), Some(currentConfig)) =>
+              /*
+               * Here, we want to check two things:
+               * - does the run should have contain a config id ?
+               *   It should if the oldest config was created a long time ago,
```

Comment:
The logic is that only nodes with initial promises send reports without a config Id. So if a node is in that case, it is because it never got genererated promises. 
If the first generated promises for that node are beyond the grace period, it means that the run should have used theses promises, and we have a (DNS) problem because it didn't.